### ISSUE TITLE
Template question

### ISSUE THREAD

**Issue by Max-Might** [transistorsoft/flutter_background_geolocation#336]
Hello.

I have a question regarding the templatig engine used.
I use a custom template.
I see that the battery level is sent as a double so instead of 99 we get 0.99.

I tried to update the template to use this: `<%= battery.level * 100 %>` but unfortunately it does not work.
What is the proper way to do this?

**Comment by christocracy** [comment]
You can't perform any functions/transformations within a template param.  You must modify your server code to perform any necessary transformations.

**Comment by Max-Might** [comment]
@christocracy Unfortunately I do not have access to the server code :(

It would be nice if the library supported a way for the user to provide a custom Location -> JSON (or map) transformer.
Something like this:

```dart
bg.BackgroundGeolocation.onLocationTransform((bg.Location event) {
  return {
    'latitude': event.coords.latitude,
    'longitude': event.coords.longitude,
    'velocity': (event.coords.speed * 100).toInt(),
  };
});
```
This way people will have much more control over how each location is transformed to JSON.

**Comment by christocracy** [comment]
@Max-Might That would be far too complex to perform 2-way communication like this over the native bridge.  Communicating between `dart` and `Objc-c/Java` is more like client/server communication.  If your `dart` code ever had an error, it could place the native code in an unusual state while it paused executing an http request awaiting your transformation object to be received.

However, there's an undocumented feature implemented in [1.9.2](https://github.com/transistorsoft/flutter_background_geolocation/blob/master/CHANGELOG.md#192---2020-07-08) allowing your to provide a native `beforeInsertBlock` for both iOS and Android.  Whatever `NSDictionary` (iOS) or `JSONObject` (Android) are returned from this block will be inserted into the plugin's SQLite database, and uploaded to your server.

By implementing these blocks for both *iOS* and *Android*, you will no longer need to define a `locationTemplate`, since it will be ignored anyway.

- [Added] Support for providing a native "beforeInsert" block in iOS `AppDelegate.m` and Android `Application.java` / `Application.kt`.  The returned object will be inserted into the SDK's SQLite database and uploaded to your `Config.url`.
__iOS__ `AppDelegate.m`
```obj-c
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
    [GeneratedPluginRegistrant registerWithRegistry:self];
    
    // [OPTIONAL] This block is called before a location is inserted into the background_geolocation SQLite database.
    // - The returned NSDictionary will be inserted.
    // - If you return nil, no record will be inserted.
    TSLocationManager *bgGeo = [TSLocationManager sharedInstance];
    bgGeo.beforeInsertBlock = ^NSDictionary *(TSLocation *tsLocation) {
        CLLocation *location = tsLocation.location;
        
        NSLog(@"[beforeInsertBlock] %@: %@", tsLocation.uuid, location);
        
        // Return a custom schema or nil to cancel SQLite insert.
        return @{
            @"lat": @(location.coordinate.latitude),
            @"lng": @(location.coordinate.longitude),
            @"battery": @{
                @"level": tsLocation.batteryLevel,
                @"is_charging": @(tsLocation.batteryIsCharging)
            }
        };
    };
    
    return [super application:application didFinishLaunchingWithOptions:launchOptions];
}
```
__Android__ `Application.java`
```java
public class Application  extends FlutterApplication {
    @Override
    public void onCreate() {        
        super.onCreate();

        BackgroundGeolocation.getInstance(this).setBeforeInsertBlock(new TSBeforeInsertBlock() {
            @Override
            public JSONObject onBeforeInsert(TSLocation tsLocation) {
                Location location = tsLocation.getLocation();
                JSONObject json = new JSONObject();
                JSONObject battery = new JSONObject();
                try {
                    json.put("lat", location.getLatitude());
                    json.put("lng", location.getLongitude());
                    
                    battery.put("level", tsLocation.getBatteryLevel());
                    battery.put("is_charging", tsLocation.getBatteryIsCharging());
                    json.put("battery", battery);
                    return json;
                } catch (JSONException e) {
                    e.printStackTrace();
                    return null;
                }
            }
        });
    }
}
```

**Comment by tspoke** [comment]
Hi @christocracy and thank you for this library.

We are currently encountering the same problem/limitation using this library. We need an integer (or null is battery == -1) server-side.

@christocracy I do not agree with you about modifying server code to match client structure/types problems. Client's libraries should never drive server/database models IMHO. Even if we can transform data before insertion on our side, adding specific code because of this library is not something I would like to do.

I like @Max-Might proposal, **providing a custom transformer is far more elegant and powerful**. Altering the data when persisted in the local database won't work in our case because we are sending a GraphQL request, not a simple body with the object in json format.

---- 

If it's too complicated to provide this kind of communication over the native bridge, why not improve the template with a full capabilities engine (methods for transforming data, ternary condition, if/else ...) ?

```
<%= battery.level * 100 %>
<%= round(battery.level * 100) %>
<%= round(battery.level) == -1 ? null : round(battery.level * 100) %>
...
```
